app/models/client.rb

```diff
@@ -28,6 +32,31 @@ def updated_grant_for(user, time_to_live)
     grant
   end
 
+  def get_redirect_uri(redirect_uri, query_params = nil, hash_params = nil)
+    unless redirect_uris && redirect_uris.split(" ").include?(redirect_uri)
+      # Wrong redirect URI, we should NOT redirect back to the client.
+      raise "Unauthorized redirect_uri: #{redirect_uri}. Requested query_params: #{query_params}, hash_params: #{hash_params}"
+    end
+    uri = URI.parse(redirect_uri)
+    if uri.fragment
+      # Note that redirect_uri is not allowed to include any fragment / hash params.
+      # Wrong redirect URI, we should NOT redirect back to the client.
+      raise "redirect_uri must not include fragment"
+    end
+    if URI.parse(APP_CONFIG[:site_url]).scheme == "https" && uri.scheme != "https"
+      # Enforce HTTPS when Portal is using HTTPS too.
+      # # Wrong redirect URI, we should NOT redirect back to the client.
+      raise "redirect_uri must use HTTPS protocol"
+    end
+    if query_params
+      redirect_uri += (redirect_uri =~ /\?/ ? "&" : "?") + URI.encode_www_form(query_params)
```

Comment:
I think there is a better way to do this, by processing the redirect_uri and then adding parameters and then returning the URI to a string again. The code above could create slightly invalid URLS if it ends with a `?` and doesn't have any params after.  I'm sure other parts of the code use this approach when adding parameters to URIs. For example the report launching code.